Framework/Core/include/Framework/TableTreeHelpers.h

```diff
@@ -22,97 +22,205 @@ namespace o2::framework
+    : ColumnToBranchBase(column, field, size)
+  {
+    if constexpr (std::is_pointer_v<T>) {
+      mLeaflist = mBranchName + "[" + std::to_string(listSize) + "]" + ROOTTypeString_t<std::remove_pointer_t<T>>;
+    } else {
+      mLeaflist = mBranchName + ROOTTypeString_t<T>;
+    }
+    mBranch = tree->GetBranch(mBranchName.c_str());
+    if (mBranch == nullptr) {
+      mBranch = tree->Branch(mBranchName.c_str(), (char*)nullptr, mLeaflist.c_str());
+    }
+    if constexpr (std::is_same_v<bool, std::remove_pointer_t<T>>) {
+      mCurrent = new bool[listSize];
+      mLast = mCurrent + listSize;
+      accessChunk(0);
+    } else {
+      accessChunk(0);
+    }
+  }
 
-  // add branches
-  bool addBranch(std::shared_ptr<arrow::ChunkedArray> col, std::shared_ptr<arrow::Field> field);
-  bool addAllBranches();
+ private:
+  std::string mLeaflist;
+
+  void resetBuffer() override
+  {
+    if constexpr (std::is_same_v<bool, std::remove_pointer_t<T>>) {
+      if (O2_BUILTIN_UNLIKELY((*mCurrentPos - mFirstIndex) * listSize >= getCurrentArray()->length())) {
+        nextChunk();
+      }
+    } else {
+      if (O2_BUILTIN_UNLIKELY((mCurrent + (*mCurrentPos - mFirstIndex) * listSize) >= mLast)) {
+        nextChunk();
+      }
+    }
+    accessChunk(*mCurrentPos);
+    mBranch->SetAddress((void*)(mCurrent + (*mCurrentPos - mFirstIndex) * listSize));
+  }
+
+  auto getCurrentArray() const
+  {
+    if (listSize > 1) {
+      return std::static_pointer_cast<o2::soa::arrow_array_for_t<std::remove_pointer_t<T>>>(std::static_pointer_cast<arrow::FixedSizeListArray>(mColumn->chunk(mCurrentChunk))->values());
+    } else {
+      return std::static_pointer_cast<o2::soa::arrow_array_for_t<std::remove_pointer_t<T>>>(mColumn->chunk(mCurrentChunk));
+    }
+  }
+
+  void nextChunk() override
+  {
+    mFirstIndex += getCurrentArray()->length();
+    ++mCurrentChunk;
+  }
+
+  void accessChunk(int at)
+  {
+    auto array = getCurrentArray();
+    if constexpr (std::is_same_v<bool, std::remove_pointer_t<T>>) {
+      for (auto i = 0; i < listSize; ++i) {
+        mCurrent[i] = (bool)array->Value((at - mFirstIndex) * listSize + i);
+      }
+    } else {
+      mCurrent = (std::remove_pointer_t<T>*)array->raw_values();
+      mLast = mCurrent + array->length() * listSize;
+    }
+  }
+
+  mutable std::remove_pointer_t<T>* mCurrent = nullptr;
+  mutable std::remove_pointer_t<T>* mLast = nullptr;
+  TBranch* mBranch = nullptr;
+};
+
+class TableToTree
+{
+ public:
+  TableToTree(std::shared_ptr<arrow::Table> table, TFile* file, const char* treename);
 
-  // write table to tree
   TTree* process();
+  void addBranch(std::shared_ptr<arrow::ChunkedArray> column, std::shared_ptr<arrow::Field> field);
+  void addAllBranches();
+
+ private:
+  arrow::Table* mTable;
+  int64_t mRows = 0;
+  TTree* mTree = nullptr;
+  std::vector<std::unique_ptr<ColumnToBranchBase>> mColumnReaders;
```

Comment:
Why do you need a unique_ptr? Are you moving them around?

Reply:
I need pointers for polymorphism so this ensures ownership.

Reply:
That's actually the main issue I have. Why do you need polimorphism at all, can't you convert the type upfront and do everything at runtime, without the need for templates?

Reply:
With too many switches, which is exactly what I wanted to avoid.

Reply:
I am all for removing the switches, but since you do have the field I think most of the if constexpr which you can become normal if, taking advantage of the field, which you have in any case. Basically, rather than having something which converts from a basic type t to root type, what you need is arrow type to root type. I see already quite a few places where that would actually work.

Reply:
Field is still dynamic, one would still need a switch to use it. Which is done once when the instance of the ColumnToBranch is added.